### PR TITLE
Move to a modern version of reflection docblock

### PR DESCRIPTION
**Pull request by UndefinedOffset**
This pull request incorporates changes from #139 but more importantly switches to phpdocumentor/reflection-docblock ^5.2 to address compatibility with PHPUnit ^9.5 that Silverstripe 4.10 supports. I believe this may also break compatibility with Silverstripe projects using PHPUnit 5.7 because of conflicting dependencies between the two versions.

### CODE REVIEW

**Comment by UndefinedOffset** [comment]
@robbieaverill hate to be a pest on this but do you know who is active as a maintainer on this repo now? This pull is a bit of a blocker if you want to use PHPUnit 9 in your project with this module.

Also #139 is a nice to have ;)

File: src/Generators/ControllerTagGenerator.php
Comment (robbieaverill) [inline src/Generators/ControllerTagGenerator.php:47]: We want to support usage on SS versions where the CMS is not installed. I guess the `isContentController` check above would stop this logic from running and causing an error if `Page` is undefined, is that right?

**Review by robbieaverill — COMMENTED** [review]
Looks good, I think the version bump would either require a minor or major release in this module

File: src/Generators/ControllerTagGenerator.php
Comment (UndefinedOffset) [reply src/Generators/ControllerTagGenerator.php:47]: Yes it would, and I just tried it on a project that does not have cms installed or any class called Page and it does seem to work without any crashes.

**Comment by UndefinedOffset** [comment]
Ya I would think it's at least a minor bump, the two versions of phpdocumentor/reflection-docblock are pretty large gaps plus the base supported PHP version is also increased (more inline with modern Silverstripe).

File: src/Generators/ControllerTagGenerator.php
Comment (Firesphere) [inline src/Generators/ControllerTagGenerator.php:51]: Instead of foreaching all the cached page classes, can you cache the `Page => Controllername` fully?
That way, instead of foreaching, you can grab the reverse of the array by Controller name, skipping the need for the foreach every time, and only have to do it once.

I minor(?) flaw here, however, is going to be pages that have the same controller, it'll be the last (or first?) found class, and all others are not included.

This, by the way, is an existing error anyway, so I don't see too much issue with it.

File: src/Generators/ControllerTagGenerator.php
Comment (Firesphere) [inline src/Generators/ControllerTagGenerator.php:14]: I _think_ this should be a protected on the Abstract

File: src/Generators/DocBlockGenerator.php
Comment (Firesphere) [inline src/Generators/DocBlockGenerator.php:115]: Can you write this as `sprintf('Class \\%s', $this->className)` please? I prefer concats to use sprintf in general, so if you see it elsewhere, please also update it :)

File: src/Generators/DocBlockGenerator.php
Comment (Firesphere) [inline src/Generators/DocBlockGenerator.php:111]: I'm not 100% sure, but `\n` could/should be written as `PHP_EOL`, just to be sure?

File: src/Generators/ControllerTagGenerator.php
Comment (Firesphere) [inline src/Generators/ControllerTagGenerator.php:55]: As I said below, can you rewrite this to a `sprintf('%s data()')`? (and same for the dataRecord)

File: src/Generators/AbstractTagGenerator.php
Comment (Firesphere) [inline src/Generators/AbstractTagGenerator.php:163]: `sprintf('@%s %s', $type, $tagstring)`

**Review by Firesphere — CHANGES_REQUESTED** [review]

File: src/Generators/DocBlockGenerator.php
Comment (UndefinedOffset) [reply src/Generators/DocBlockGenerator.php:111]: Only thing with `PHP_EOL` is on windows that resolves to `\r\n`, which PSR-2 and PSR-12 doesn't allow. That said ideally it should detect the files major EOL I would think and use that. I'll see if PHPDocumentor has anything for this or if we can come up with something quick for the detection.

**Comment by UndefinedOffset** [comment]
I've pushed some changes, see above for the one outstanding issue

File: src/Generators/DocBlockGenerator.php
Comment (Firesphere) [reply src/Generators/DocBlockGenerator.php:111]: Hmmm, fair call.

Problem however, would be that some IDE's screw this up, and go with `\r\n` instead...

I agree, use `"\n"` and document the caveat for Windows users

**Comment by UndefinedOffset** [comment]
Note about `\n` on windows was added to the bottom of the readme in the "A word of caution" section.

**Review by Firesphere — APPROVED** [review]

**Comment by Firesphere** [comment]
LGTM now.
@robbieaverill Any thoughts?

I want to give it a few manual tests before merging and tagging a new release. I'm guessing this would be release line 4.x, given the updates.

**Comment by robbieaverill** [comment]
LGTM, it might be safest as a new major version release yeah.

**Comment by Firesphere** [comment]
A quick test shows all classes being prefixed with a `\`:
```
* @property string $Name
 * @property string $Lat
 * @property string $Lng
 * @method \DataList|\Location[] Locations()
 * @method \DataList|\Suburb[] Suburbs()
 */
class City extends DataObject
```

That does not sound correct to me

File: src/Generators/DocBlockGenerator.php
Comment (Firesphere) [inline src/Generators/DocBlockGenerator.php:115]: It looks like it might be here, something goes wrong and all classes are prefixed with a `\`

**Review by Firesphere — CHANGES_REQUESTED** [review]
All `@method ` tag classes are prefixed with a `\`, this causes the class paths to be wrong.

**Comment by UndefinedOffset** [comment]
@Firesphere this appears to be the behavior in the current version as well, and technically would make sense I think otherwise it would be relative to the current namespace of the file (at least if you tried to use that same path in PHP). VSCode using intellephence does seem to properly handle this, I'm curious how does your ide of choice handle it?

**Comment by Firesphere** [comment]
PHPStorm does not handle this properly, and I've not seen it do that before, as it shouldn't include the namespace, as far as I remember...

**Comment by UndefinedOffset** [comment]
Hmm I think that may cause more issues, particularly when it comes to aliased classes in the use statement. It looks like at least based on the PHPDocumentor docs that class names in types can be the FQCN or the local, but to do the local we'd need to parse the document I would think to figure out the alias if present as a use statement.

**Comment by Firesphere** [comment]
This has been discussed in a different PR, where we decided against both, IIRC.
Not adding the class as a `use` is the safer way to go.

**Comment by UndefinedOffset** [comment]
@Firesphere I think the issue you are seeing is because it's not resolving to the full path to the class. It really should be `\SilverStripe\ORM\DataList` not `\DataObject` do you have the composer.json, composer.lock and sample code you used so I can see if I can re-produce what happened there?

**Comment by Firesphere** [comment]
Yes, it's literally above, you could probably determine my DB/Relations from the example there :)

**Comment by UndefinedOffset** [comment]
hmmm I didn't not observe that behavior using a model I have it generated as I would expect. Is it possible PHPStorm is adjusting them after this branch generates them?

```
/**
 * Class \Project\Model\Elements\CarouselElement
 *
 * @property int $SlideDelay
 * @property bool $DisableCarouselDots
 * @method \SilverStripe\ORM\DataList|\Project\Model\Elements\Children\CarouselElementItem[] Images()
 */
```

My local composer.lock has the following for PHPDocumentor's dependencies:

- phpdocumentor/reflection-common 2.2.0
- phpdocumentor/reflection-docblock 5.3.0
- phpdocumentor/type-resolver 1.6.1

Perhaps there is a miss-match in dependencies as well that maybe causing the problem?

**Comment by Firesphere** [comment]
Right, I see the problem, I think.

If the class is already in the `use` statements somewhere, it's being shorted, with the preceding `\`

**Comment by UndefinedOffset** [comment]
Ah do you have the `SilverLeague\IDEAnnotator\DataObjectAnnotator.use_short_name` config option set to true by chance? If so I think we need to add a condition around the lines you mentioned in `DocBlockGenerator.php`.

**Comment by UndefinedOffset** [comment]
If so I believe that's the source of the issue, I'm not sure that will work properly with the newer versions of phpDocumentor.

**Comment by Firesphere** [comment]
We can't remove that option, sadly/probably.

Reason is, that if people go in to the file, and import the class with a use case, the IDE like PHPStorm, will shorten it.

A build will then lengthen it again, or break it. Neither of those is a desirable outcome.

**Comment by UndefinedOffset** [comment]
@Firesphere sorry for the slow follow up I've added a new resolver that seems to address the issue you called out. Let me know 😄

**Comment by Firesphere** [comment]
While at it, could you please update the following?
`AbstractTagGenerator.php::116`
`        return (array)call_user_func_array('array_merge', array_values($tags));`

This will make it PHP 8.1 compatible in a single go

**Comment by Firesphere** [comment]
I have no other comments, but would like that minor change to be added in this PR and make it all go to a new point release

**Comment by UndefinedOffset** [comment]
I've replaced the line you referenced with your line @Firesphere (with `$this->tags` instead of `$tags` 😉)

**Review by Firesphere — APPROVED** [review]

**Pull request merged**
